### PR TITLE
feat(subcommands): implement per-subcommand preconditions

### PR DESCRIPTION
**Pull request by favna**
Also took a moment to update the tsconfig a bit and add some more tsdoc. I'll go through the tsconfig of the other packages in another PR after this to ensure proper conventional commit scoping.

### CODE REVIEW

**Comment by favna** [comment]
@sapphiredev pack

**Comment by favna** [comment]
Branch that was used for runtime testing: https://github.com/favna/musical-giggle/pull/139

File: packages/subcommands/src/lib/Subcommand.ts [hunk 1]
```diff
@@ -249,14 +322,54 @@ export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand
...
 			command: this,
 			context,
 			matchedSubcommandMapping: subcommand
 		};
 
+		// Check if any subcommand preconditions were defined for thus subcommand:
+		const preconditionsForSubcommand = this.subcommandPreconditions.get(
+			subcommandGroupName ? `${subcommandGroupName}.${subcommand.name}` : subcommand.name
+		);
+		if (preconditionsForSubcommand) {
```
Comment (vladfrangu) [inline packages/subcommands/src/lib/Subcommand.ts:357]:
```suggestion
		);

		if (preconditionsForSubcommand) {
```

File: packages/subcommands/src/lib/Subcommand.ts [hunk 2]
```diff
@@ -249,14 +322,54 @@ export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand
...
+		args.restore();
+
+		const params: Partial<Pick<MessageCommandDeniedPayload, 'parameters'>> = {};
+
+		if (parameters.isOk()) {
+			params.parameters = parameters.unwrap();
+		}
+
+		return params;
+	}
```
Comment (vladfrangu) [inline packages/subcommands/src/lib/Subcommand.ts:337]: What exactly is the point of this?

File: packages/subcommands/src/lib/Subcommand.ts [hunk 3]
```diff
@@ -300,6 +414,19 @@ export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand
 			matchedSubcommandMapping: subcommand
 		};
 
+		// Check if any subcommand preconditions were defined for thus subcommand:
+		const preconditionsForSubcommand = this.subcommandPreconditions.get(
+			subcommandGroupName ? `${subcommandGroupName}.${subcommand.name}` : subcommand.name
+		);
+		if (preconditionsForSubcommand) {
```
Comment (vladfrangu) [inline packages/subcommands/src/lib/Subcommand.ts:421]:
```suggestion
		);

		if (preconditionsForSubcommand) {
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/subcommands/src/lib/Subcommand.ts (hunk 2 above)
Comment (favna) [reply packages/subcommands/src/lib/Subcommand.ts:337]: Mimicking the payload that normal preconditions get

See https://github.com/sapphiredev/framework/blob/9103253c8da170c229dd87f5e22ef0031d96c436/src/optional-listeners/message-command-listeners/CorePreMessageCommandRun.ts#L9-L17

Which gets called from: https://github.com/sapphiredev/framework/blob/9103253c8da170c229dd87f5e22ef0031d96c436/src/optional-listeners/message-command-listeners/CorePrefixedMessage.ts#L40-L46

And it is also in the interface: https://github.com/sapphiredev/framework/blob/9103253c8da170c229dd87f5e22ef0031d96c436/src/lib/types/Events.ts#L455-L458

File: packages/subcommands/src/lib/Subcommand.ts [hunk 4]
```diff
@@ -14,8 +24,32 @@ import {
...
+ * The reason for this is that this class implements {@link Command.messageRun} and {@link Command.chatInputRun}
+ * which you wouldn't necessarily want to do for commands that don't have subcommands.
+ * Furthermore, it also just adds general unnecessary memory overhead.
+ */
 export class Subcommand<PreParseReturn extends Args = Args, O extends Subcommand.Options = Subcommand.Options> extends Command<PreParseReturn, O> {
+	/**
+	 * The preconditions to be run for each specified subcommand.
+	 * @since 4.1.0
+	 */
+	public subcommandPreconditions = new Map<string, PreconditionContainerArray>();
```
Comment (vladfrangu) [inline packages/subcommands/src/lib/Subcommand.ts:41]:
```suggestion
	public readonly subcommandPreconditions = new Map<string, PreconditionContainerArray>();
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**